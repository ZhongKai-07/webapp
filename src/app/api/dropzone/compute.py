import os
import tempfile
import osmnx as ox  # For OpenStreetMap data
import geopandas as gpd  # For handling geospatial data
import pandas as pd  # For data manipulation
from shapely.geometry import MultiPolygon
from osmnx._errors import InsufficientResponseError
from shapely.geometry import shape
import shutil
from shapely.validation import make_valid  # To fix invalid geometries
from shapely.ops import unary_union  # For geometric operations
from shapely import to_geojson
import json
import random
from shapely.geometry import Point, Polygon, MultiPolygon
from shapely.geometry import shape
import geopandas as gpd
from shapely import to_geojson
import glob

# Initialize a global variable for bounding box (shape of the district/area you are loading)
bbox = None

dropzone_feature_type_property_name = "dropzone_feature_type"

path_prefix = "."



def prepare_bbox(district_shape):
    if (
        district_shape["type"] == "Feature"
        and district_shape["geometry"]["type"] == "Polygon"
    ):
        coordinates = district_shape["geometry"]["coordinates"]
        if isinstance(coordinates[0][0], list):
            # Flatten the coordinates if they are nested
            coordinates = [coord for sublist in coordinates for coord in sublist]
        district_shapely = Polygon(coordinates)
    elif (
        "geometry" in district_shape
        and district_shape["geometry"]["type"] == "MultiPolygon"
    ):
        district_shapely = MultiPolygon(
            [
                Polygon(poly[0], poly[1:])
                for poly in district_shape["geometry"]["coordinates"]
            ]
        )
    elif district_shape["type"] == "FeatureCollection":
        district_shapely = unary_union(
            [
                shape(feature["geometry"])
                if shape(feature["geometry"]).is_valid
                else make_valid(shape(feature["geometry"]))
                for feature in district_shape["features"]
            ]
        )
        if isinstance(district_shapely, Polygon):
            district_shapely = MultiPolygon([district_shapely])

    polygon = shape(district_shapely)

    # Validate and possibly fix the geometry
    if not polygon.is_valid:
        polygon = make_valid(polygon)

    # Bounds of the district shape
    global bbox
    bbox = polygon
    return polygon


# Function to extract OSM features within a bounding polygon
def extract_osm_features(gdf, feature_type, exclusion_tags=None):
    try:
        # Query OSM for the specified feature type within the bounding box
        if "=" in feature_type:
            feature_name, feature_value = feature_type.split("=")
            gdf_features = ox.features_from_polygon(
                bbox, tags={feature_name: feature_value}
            )
        else:
            gdf_features = ox.features_from_polygon(bbox, tags={feature_type: True})

        # Check if the returned data is a valid GeoDataFrame
        if (
            not isinstance(gdf_features, gpd.GeoDataFrame)
            or "geometry" not in gdf_features.columns
        ):
            print(
                f"'gdf_features' is not a GeoDataFrame or does not have a 'geometry' column."
            )
            return gpd.GeoDataFrame()

        # Filter out points from the GeoDataFrame
        gdf_features = gdf_features[gdf_features.geometry.type != "Point"]

        # Filter out features based on exclusion tags
        if exclusion_tags:
            for key, value in exclusion_tags.items():
                if key in gdf_features.columns:
                    gdf_features = gdf_features[
                        ~gdf_features[key].astype(str).eq(str(value))
                    ]

        # Add the feature type as a column
        gdf_features[dropzone_feature_type_property_name] = [feature_type] * len(
            gdf_features
        )
        print(f"Found {len(gdf_features)} {feature_type} features")
        return gdf_features

    except InsufficientResponseError:
        # Handle exception when no data is found for the feature type
        print(f"No data found for {feature_type} in the specified area.")
        return gpd.GeoDataFrame()


# Function to clean data types in GeoDataFrame
def clean_gdf(gdf):
    # Convert object types to strings in the GeoDataFrame
    for column in gdf.columns:
        if gdf[column].dtype == "object":
            gdf[column] = gdf[column].apply(
                lambda x: str(x) if isinstance(x, list) else x
            )
    return gdf


# Main function to process OSM data for a given district
def compute_osm():
    # Read the GeoJSON file into a GeoDataFrame
    gdf = gpd.GeoDataFrame(geometry=[bbox])

    # Concatenate features from various OSM categories
    all_features = gpd.GeoDataFrame(
        pd.concat(
            [
                # Extracting different types of features from OSM
                extract_osm_features(
                    gdf,
                    "highway",
                    exclusion_tags={"tunnel": "yes", "area:highway": "pedestrian"},
                ),
                extract_osm_features(gdf, "landuse=forest"),
                extract_osm_features(gdf, "natural=wood"),
                extract_osm_features(gdf, "historic=*"),
                extract_osm_features(gdf, "natural=hill"),
                extract_osm_features(gdf, "natural=peak"),
                extract_osm_features(
                    gdf, "landuse=railway", exclusion_tags={"tunnel": "yes"}
                ),
                extract_osm_features(gdf, "railway", exclusion_tags={"tunnel": "yes"}),
                extract_osm_features(gdf, "boundary=forest_compartment"),
                extract_osm_features(gdf, "boundary=forest"),
                extract_osm_features(gdf, "aeroway"),
                extract_osm_features(gdf, "landuse=cemetery"),
                extract_osm_features(gdf, "landuse=quarry"),
                extract_osm_features(gdf, "man_made=mast"),
                extract_osm_features(gdf, "natural=cliff"),
                extract_osm_features(gdf, "natural=water"),
                extract_osm_features(gdf, "power=plant"),
                extract_osm_features(gdf, "leisure=golf_course"),
                extract_osm_features(gdf, "leisure=park"),
                extract_osm_features(gdf, "leisure=stadium"),
                extract_osm_features(gdf, "leisure=nature_reserve"),
                extract_osm_features(gdf, "building=church"),
                extract_osm_features(gdf, "amenity=fountain"),
                extract_osm_features(gdf, "natural=scrub"),
                extract_osm_features(gdf, "waterway"),
                extract_osm_features(gdf, "designation=national_park"),
            ],
            ignore_index=True,
        )
    )
    print("Clean GDF")
    # Clean the concatenated GeoDataFrame
    all_features_cleaned = clean_gdf(all_features)

    # Clip the cleaned features to the bounding box (some roads lead outside the district)
    all_features_cleaned_clipped = gpd.clip(gdf=all_features_cleaned, mask=bbox)

    # Perform unary union operation on the clipped geometries
    # I removed this now to retain the feature types
    # gdf_conv = unary_union(all_features_cleaned_clipped["geometry"])

    # Modify the GeoDataFrame to contain only the clipped geometries and the dropzone_feature_type_property_name column
    geodataframe = all_features_cleaned_clipped[
        ["geometry", dropzone_feature_type_property_name]
    ]

    os.makedirs(f"{path_prefix}/{district}/excluded", exist_ok=True)

    with open(
        f"{path_prefix}/{district}/excluded/exclusions.json", "w"
    ) as output_file:
        output_file.write(geodataframe.to_json())
    return all_features_cleaned_clipped


def random_point_in_bounds(bounds):
    minx, miny, maxx, maxy = bounds
    return (random.uniform(minx, maxx), random.uniform(miny, maxy))


def create_square_around_point(point, size=5):
    x, y = point
    delta_deg = size / 111320  # Convert 5 meters to degrees
    square_coords = [
        [x - delta_deg, y - delta_deg],
        [x + delta_deg, y - delta_deg],
        [x + delta_deg, y + delta_deg],
        [x - delta_deg, y + delta_deg],
        [x - delta_deg, y - delta_deg],
    ]
    return square_coords


def is_square_within_polygon(square_coords, polygon):
    for coord in square_coords[:-1]:
        if not polygon.contains(Point(coord)):
            return False
    return True


def create_dropzones(district, district_shape, exclusionzones):
    # Convert district shape to a Shapely geometry

    district_shapely = prepare_bbox(district_shape)

    random_squares_set = []

    while len(random_squares_set) < max_squares:
        # Generate a random square within the bounds
        point = random_point_in_bounds(district_shapely.bounds)
        square_coords = create_square_around_point(point)
        square = Polygon(square_coords)

        # Check if the square is within the district and does not intersect with any exclusion zones
        if (
            district_shapely.contains(square)
            and not exclusionzones.intersects(square).any()
        ):
            random_squares_set.append(square)

        percent_progress = (len(random_squares_set) / max_squares) * 100
        if percent_progress % 5 == 0 and percent_progress // 5 != 0:
            print(round(percent_progress, 2), "%")

    random_squares_gdf = gpd.GeoDataFrame(geometry=random_squares_set)

    all_features_cleaned_clipped = gpd.clip(random_squares_gdf, district_shapely)

    os.makedirs(f"{path_prefix}/{district}", exist_ok=True)

    with open(f"{path_prefix}/{district}/dropzones.json", "w") as output_file:
        output_file.write(all_features_cleaned_clipped.to_json())


# Main execution
if __name__ == "__main__":
    # Load districts from all files with .json extension
    for district_file in glob.glob("*.json"):
        district = district_file.split(".")[0]

        # Generate folders for district and excluded
        district_folder = os.path.join(os.getcwd(), district)
        excluded_folder = os.path.join(district_folder, "excluded")

        os.makedirs(district_folder, exist_ok=True)
        os.makedirs(excluded_folder, exist_ok=True)

        with open(f"{district_file}", "r") as output_file:
            district_shape = prepare_bbox(json.load(output_file))

        max_squares = 750
        print(f"==================== {district} ====================")
        with open(district_file, "r") as file:
            district_shape = json.load(file)
        print(f"==================== Loading Exclusion Zones ====================")
        exclusionzones = compute_osm()
        print(f"==================== Calculating Drop Zones ====================")
        create_dropzones(district, district_shape, exclusionzones)
