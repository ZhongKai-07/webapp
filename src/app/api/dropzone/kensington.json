{
  "type": "Feature",
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [-0.1560504330356739, 51.497513058659756],
        [-0.1578770102604991, 51.49902006721774],
        [-0.15847537176518323, 51.50225030730073],
        [-0.16102628133778396, 51.50158853397743],
        [-0.16260091687642642, 51.50029119617535],
        [-0.16550611944522167, 51.49869900887278],
        [-0.16580530019756373, 51.49924284239588],
        [-0.16764762377777537, 51.49829932399436],
        [-0.16841132201401693, 51.49902006721774],
        [-0.1735682534030709, 51.498220697460894],
        [-0.1795676148052985, 51.497775147104626],
        [-0.1803391862192333, 51.50146404196612],
        [-0.18426002871045294, 51.50179165252221],
        [-0.187936802693183, 51.510178482757944],
        [-0.1921410795813583, 51.50987052883523],
        [-0.19298350959453198, 51.511816535538365],
        [-0.19406213493850202, 51.51175756563827],
        [-0.19506990168323318, 51.515033671199106],
        [-0.19921119314986277, 51.5145488075761],
        [-0.1998961596091722, 51.51677655935747],
        [-0.20074646280003913, 51.51797561399273],
        [-0.20148654150320106, 51.517798704292446],
        [-0.20376976303423258, 51.52062270728589],
        [-0.20266751815718287, 51.520897900152995],
        [-0.20063623831233415, 51.52070788603047],
        [-0.2016361318793721, 51.522739071478185],
        [-0.20387998752193753, 51.52345981470157],
        [-0.20617895540835549, 51.52594310271668],
        [-0.210280880986519, 51.526683502573434],
        [-0.21331205439840567, 51.526899725540446],
        [-0.21557165639635756, 51.52672281584016],
        [-0.21602830070256385, 51.52792842268655],
        [-0.21798084877048046, 51.528642613698814],
        [-0.22363379035420677, 51.52992684707866],
        [-0.22544462122364556, 51.53016927889016],
        [-0.227688476866211, 51.52992029486754],
        [-0.22758612555619925, 51.53015617446792],
        [-0.22849941416861186, 51.53035274080157],
        [-0.228727736321715, 51.529880981600805],
        [-0.22741291564694857, 51.52788910941982],
        [-0.22719246667153864, 51.52478991355927],
        [-0.22661772469993416, 51.52479646577039],
        [-0.2267279491876391, 51.52197901498808],
        [-0.22830258472628154, 51.521094466486645],
        [-0.22260240407639598, 51.515643026833416],
        [-0.21928779626755368, 51.50957567933475],
        [-0.2178863706381619, 51.50975914124616],
        [-0.21805170736971938, 51.51023090044692],
        [-0.21720927735654566, 51.51039470572496],
        [-0.21683923800496469, 51.50944463511232],
        [-0.21581572490484713, 51.50911702455623],
        [-0.21520161704477658, 51.50764932926498],
        [-0.21614639836796204, 51.50751828504254],
        [-0.21596531528101814, 51.50640440915186],
        [-0.21798872194817367, 51.50614887291812],
        [-0.21704394062498822, 51.50449771571546],
        [-0.2130601127122229, 51.49984564581907],
        [-0.2102730078088258, 51.49812896650519],
        [-0.20799765945548748, 51.496019154524014],
        [-0.20739142477311015, 51.49615019874645],
        [-0.20302968433107063, 51.49331964354189],
        [-0.2018329613217024, 51.491851948250634],
        [-0.20238408376022723, 51.49166848633923],
        [-0.19821917276051804, 51.48806477022231],
        [-0.1964477077795453, 51.487206430565365],
        [-0.19015703880266882, 51.48951936109132],
        [-0.1911962982581728, 51.490810146682286],
        [-0.18857453008633318, 51.49164227749474],
        [-0.1879919149370355, 51.49092153427136],
        [-0.18665347472918942, 51.49130811472753],
        [-0.18607873275758494, 51.49065289361537],
        [-0.18479540479359136, 51.49094119090472],
        [-0.1831262911226304, 51.49077738562668],
        [-0.18126822118703234, 51.488726543545596],
        [-0.17912671685447865, 51.48710159518742],
        [-0.17180466159979135, 51.49185850046175],
        [-0.17011192839575076, 51.490253208736945],
        [-0.16978912811032906, 51.49033838748153],
        [-0.16849005379094906, 51.4922057676512],
        [-0.16972614268878336, 51.493031346252536],
        [-0.16418342559276203, 51.49654988362487],
        [-0.1634512200672933, 51.49626813854664],
        [-0.16290009762876845, 51.49678576322525],
        [-0.1560504330356739, 51.497513058659756]
      ]
    ]
  },
  "properties": {
    "PCON13CD": "E14000768",
    "PCON13CDO": "C41",
    "PCON13NM": "Kensington"
  },
  "id": "E14000768"
}
