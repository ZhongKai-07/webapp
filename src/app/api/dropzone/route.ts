import fs from 'fs/promises';
import { Feature, FeatureCollection } from 'geojson';
import { NextResponse } from 'next/server';
import path from 'path';

const data: Record<string, FeatureCollection | Feature> = {};
const directoryPath = 'src/app/api/dropzone';

export async function GET() {
  // Scan the directory for all json files and then load them and put them in an object based on their filename without extension
  const filesInDir = await fs.readdir(directoryPath);
  await Promise.all(filesInDir.map(scanDirectory));
  return NextResponse.json(data);
}
async function scanDirectory(
  file: string
): Promise<Record<string, FeatureCollection | Feature>> {
  const filePath = path.join(directoryPath, file);
  const fileStats = await fs.stat(filePath);

  if (fileStats.isFile() && path.extname(file) === '.json') {
    const fileName = path.basename(file, '.json');
    const fileData = await fs.readFile(filePath, 'utf-8');
    const jsonData = JSON.parse(fileData);

    data[fileName] = jsonData;
  }

  return data;
}
