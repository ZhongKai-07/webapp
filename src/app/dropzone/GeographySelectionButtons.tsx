import {
  GeographySelection,
  PredefinedSelection,
} from '@/app/dropzone/GeographySelection';

interface GeographySelectionButtonsProps {
  onIndividualButtonClick: () => void;
  onPredefinedSelectionButtonClick: (selection: string) => void;
  geographySelection: GeographySelection;
  renderPreSelected: boolean;
}

export const GeographySelectionButtons: React.FC<
  GeographySelectionButtonsProps
> = ({
  onIndividualButtonClick,
  onPredefinedSelectionButtonClick,
  geographySelection,
  renderPreSelected,
}) => {
  return (
    <div className='w-1/3justify-center flex gap-[1em] mt-[1em]'>
      <button
        className={`rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow ${
          geographySelection === GeographySelection.Individual
            ? 'bg-apian-yellow'
            : 'bg-white hover:bg-apian-yellow-opacity-50'
        }`}
        onClick={() => onIndividualButtonClick()}
      >
        Individual
      </button>
      {renderPreSelected &&
        Object.entries(PredefinedSelection).map((selection) => (
          <button
            className={`rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow ${
              geographySelection ===
              (selection[1] as unknown as GeographySelection)
                ? 'bg-apian-yellow'
                : 'bg-white hover:bg-apian-yellow-opacity-50'
            }`}
            onClick={() => onPredefinedSelectionButtonClick(selection[1])}
            key={selection[0]}
          >
            {selection[1].charAt(0).toUpperCase() + selection[1].slice(1)}
          </button>
        ))}
    </div>
  );
};
