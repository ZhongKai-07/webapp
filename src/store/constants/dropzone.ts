import type { constant } from '@/store/types';

export enum DROPZONE_ERROR {
  'DROPZONE_NO_ERROR' = 'DROPZONE_NO_ERROR',
  'DROPZONE_DATA_FETCH_ERROR' = 'DROPZONE_DATA_FETCH_ERROR',
  'DROPZONE_OUTLINE_FETCH_ERROR' = 'DROPZONE_OUTLINE_FETCH_ERROR',
}

export const DROPZONE_SET_SHOULD_SHOW_MAP_CONTROLS: constant =
  'DROPZONE_SET_SHOULD_SHOW_MAP_CONTROLS';
export const DROPZONE_SET_GEOGRAPHY_SELECTION: constant =
  'DROPZONE_SET_GEOGRAPHY_SELECTION';
export const DROPZONE_SET_VISIBILITIES: constant = 'DROPZONE_SET_VISIBILITIES';
export const DROPZONE_CONFIRM_SELECTION: constant =
  'DROPZONE_CONFIRM_SELECTION';
export const DROPZONE_SET_COORDINATES: constant = 'DROPZONE_SET_COORDINATES';
export const DROPZONE_SET_SELECTED_AREA: constant =
  'DROPZONE_SET_SELECTED_AREA';
export const DROPZONE_SET_EXCLUSIONS: constant = 'DROPZONE_SET_EXCLUSIONS';
export const DROPZONE_SET_ERROR: constant = 'DROPZONE_SET_ERROR';
export const DROPZONE_MAP_LOADING: constant = 'DROPZONE_MAP_LOADING';
export const DROPZONE_SET_OUTLINE: constant = 'DROPZONE_SET_OUTLINE';
export const DROPZONE_LOAD_OUTLINE: constant = 'DROPZONE_LOAD_OUTLINE';
