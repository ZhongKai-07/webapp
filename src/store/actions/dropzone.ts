import { createAction } from '@reduxjs/toolkit';
import {
  Feature,
  FeatureCollection,
  MultiPolygon,
  Point,
  Polygon,
} from 'geojson';

import type { DROPZONE_ERROR } from '@/store/constants/dropzone';
import {
  DROPZONE_CONFIRM_SELECTION,
  DROPZONE_LOAD_OUTLINE,
  DROPZONE_MAP_LOADING,
  DROPZONE_SET_COORDINATES,
  DROPZONE_SET_ERROR,
  DROPZONE_SET_EXCLUSIONS,
  DROPZONE_SET_GEOGRAPHY_SELECTION,
  DROPZONE_SET_OUTLINE,
  DROPZONE_SET_SELECTED_AREA,
  DROPZONE_SET_SHOULD_SHOW_MAP_CONTROLS,
  DROPZONE_SET_VISIBILITIES,
} from '@/store/constants/dropzone';

import { GeographySelection } from '@/app/dropzone/GeographySelection';

export const dropZoneSetGeographySelection = createAction<GeographySelection>(
  DROPZONE_SET_GEOGRAPHY_SELECTION
);

export const dropZoneSetShouldShowMapControls = createAction<boolean>(
  DROPZONE_SET_SHOULD_SHOW_MAP_CONTROLS
);

export const dropZoneSetVisibilities = createAction<{
  markers: boolean;
  exclusions: boolean;
}>(DROPZONE_SET_VISIBILITIES);

export const dropZoneConfirmSelection = createAction<GeographySelection>(
  DROPZONE_CONFIRM_SELECTION
);

export const dropZoneSetSelectedArea = createAction<
  FeatureCollection<Point | Polygon> | undefined
>(DROPZONE_SET_SELECTED_AREA);

export const dropZoneSetExclusions = createAction<
  FeatureCollection<Polygon> | undefined
>(DROPZONE_SET_EXCLUSIONS);

export const dropZoneSetCoordinates = createAction<Polygon | undefined>(
  DROPZONE_SET_COORDINATES
);

export const dropZoneSetOutline =
  createAction<
    Record<
      string,
      | FeatureCollection<Polygon | MultiPolygon>
      | Feature<Polygon | MultiPolygon>
    >
  >(DROPZONE_SET_OUTLINE);

export const dropZoneLoadOutline = createAction(DROPZONE_LOAD_OUTLINE);

export const dropZoneSetError =
  createAction<DROPZONE_ERROR>(DROPZONE_SET_ERROR);

export const dropZoneSetMapLoading =
  createAction<boolean>(DROPZONE_MAP_LOADING);
