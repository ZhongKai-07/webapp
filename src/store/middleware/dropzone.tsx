import { Action, MiddlewareAPI } from '@reduxjs/toolkit';
import centerOfMass from '@turf/center-of-mass';
import circle from '@turf/circle';
import { feature, featureCollection } from '@turf/helpers';
import axios from 'axios';

import {
  DROPZONE_CONFIRM_SELECTION,
  DROPZONE_ERROR,
  DROPZONE_LOAD_OUTLINE,
} from '@/store/constants/dropzone';

import { GeographySelection } from '@/app/dropzone/GeographySelection';

import {
  dropZoneSetCoordinates,
  dropZoneSetError,
  dropZoneSetExclusions,
  dropZoneSetGeographySelection,
  dropZoneSetMapLoading,
  dropZoneSetOutline,
  dropZoneSetSelectedArea,
} from '../actions/dropzone';

const DROPZONE_API = process.env.NEXT_PUBLIC_DROPZONE_API as string;

const dropZoneMiddleware =
  (api: MiddlewareAPI) =>
  (next: (action: Action) => void) =>
  (action: Action) => {
    next(action);

    if (action.type === DROPZONE_CONFIRM_SELECTION) {
      api.dispatch(dropZoneSetMapLoading(true));
      const { geographySelection, coordinates } = api.getState().dropZone;

      api.dispatch(dropZoneSetExclusions(undefined));
      api.dispatch(dropZoneSetSelectedArea(undefined));
      api.dispatch(dropZoneSetCoordinates(undefined));

      if (geographySelection === GeographySelection.Individual && coordinates) {
        const polygonFeature = feature(coordinates);
        api.dispatch(
          dropZoneSetSelectedArea(featureCollection([polygonFeature]))
        );
        api.dispatch(
          dropZoneSetExclusions(
            featureCollection([
              circle(centerOfMass(polygonFeature), 0.5, {
                units: 'kilometers',
              }),
            ])
          )
        );
        api.dispatch(dropZoneSetCoordinates(undefined));
      } else {
        axios
          .get(`${DROPZONE_API}/${geographySelection}`)
          .then((res) => {
            const featureData = res.data;
            api.dispatch(dropZoneSetError(DROPZONE_ERROR.DROPZONE_NO_ERROR));
            api.dispatch(dropZoneSetSelectedArea(featureData));
          })
          .catch(() => {
            api.dispatch(
              dropZoneSetError(DROPZONE_ERROR.DROPZONE_DATA_FETCH_ERROR)
            );
          });

        axios
          .get(`${DROPZONE_API}/${geographySelection}/excluded`)
          .then((res) => {
            const featureData = res.data;
            api.dispatch(dropZoneSetExclusions(featureData));
            api.dispatch(dropZoneSetError(DROPZONE_ERROR.DROPZONE_NO_ERROR));
          })
          .catch(() => {
            api.dispatch(
              dropZoneSetError(DROPZONE_ERROR.DROPZONE_DATA_FETCH_ERROR)
            );
          });

        api.dispatch(dropZoneSetCoordinates(coordinates));
      }

      api.dispatch(
        dropZoneSetGeographySelection(GeographySelection.Display_Results)
      );
      api.dispatch(dropZoneSetMapLoading(false));
    } else if (action.type === DROPZONE_LOAD_OUTLINE) {
      axios
        .get(DROPZONE_API)
        .then((res) => {
          const outlineData = res.data;
          api.dispatch(dropZoneSetError(DROPZONE_ERROR.DROPZONE_NO_ERROR));
          api.dispatch(dropZoneSetOutline(outlineData));
        })
        .catch(() => {
          api.dispatch(
            dropZoneSetError(DROPZONE_ERROR.DROPZONE_OUTLINE_FETCH_ERROR)
          );
        });
    }
  };

export default dropZoneMiddleware;
