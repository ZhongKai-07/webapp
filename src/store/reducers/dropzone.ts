import { createReducer } from '@reduxjs/toolkit';

import {
  dropZoneConfirmSelection,
  dropZoneSetError,
  dropZoneSetExclusions,
  dropZoneSetGeographySelection,
  dropZoneSetMapLoading,
  dropZoneSetOutline,
  dropZoneSetSelectedArea,
  dropZoneSetShouldShowMapControls,
  dropZoneSetVisibilities,
} from '@/store/actions/dropzone';
import { DROPZONE_ERROR } from '@/store/constants/dropzone';
import { DropZoneState } from '@/store/types';

import { GeographySelection } from '@/app/dropzone/GeographySelection';

const initialState: DropZoneState = {
  shouldShowMapControls: false,
  geographySelection: GeographySelection.Individual,
  visibilities: { markers: true, exclusions: false },
  markers: undefined,
  exclusions: undefined,
  coordinates: undefined,
  error: DROPZONE_ERROR.DROPZONE_NO_ERROR,
  selectedArea: undefined,
  mapLoading: false,
};

const dropZoneReducer = createReducer(initialState, (builder) => {
  builder
    .addCase(dropZoneSetGeographySelection, (state, action) => {
      state.geographySelection = action.payload;
      state.visibilities = { markers: true, exclusions: false };
      state.shouldShowMapControls = false;
      state.error = DROPZONE_ERROR.DROPZONE_NO_ERROR;
    })
    .addCase(dropZoneSetVisibilities, (state, action) => {
      state.visibilities = action.payload;
    })
    .addCase(dropZoneSetShouldShowMapControls, (state, action) => {
      state.shouldShowMapControls = action.payload;
    })
    .addCase(dropZoneConfirmSelection, (state, action) => {
      state.geographySelection = action.payload;
    })
    .addCase(dropZoneSetSelectedArea, (state, action) => {
      state.selectedArea = action.payload;
    })
    .addCase(dropZoneSetExclusions, (state, action) => {
      state.exclusions = action.payload;
    })
    .addCase(dropZoneSetMapLoading, (state, action) => {
      state.mapLoading = action.payload;
    })
    .addCase(dropZoneSetError, (state, action) => {
      state.error = action.payload;
    })
    .addCase(dropZoneSetOutline, (state, action) => {
      state.outlineData = action.payload;
    })
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    .addDefaultCase(() => {});
});

export default dropZoneReducer;
