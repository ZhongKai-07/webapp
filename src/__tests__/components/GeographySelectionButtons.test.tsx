import { cleanup, render, screen } from '@testing-library/react';

import {
  GeographySelection,
  PredefinedSelection,
} from '@/app/dropzone/GeographySelection';
import { GeographySelectionButtons } from '@/app/dropzone/GeographySelectionButtons';

describe('GeographySelectionButtons', () => {
  afterEach(cleanup);
  it('should match the snapshot with preselected', () => {
    const { container } = render(
      <GeographySelectionButtons
        onIndividualButtonClick={jest.fn()}
        onPredefinedSelectionButtonClick={jest.fn()}
        geographySelection={GeographySelection.Individual}
        renderPreSelected={true}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot without preselected', () => {
    const { container } = render(
      <GeographySelectionButtons
        onIndividualButtonClick={jest.fn()}
        onPredefinedSelectionButtonClick={jest.fn()}
        geographySelection={GeographySelection.Individual}
        renderPreSelected={false}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should call onIndividualButtonClick when individual button is clicked', () => {
    const onIndividualButtonClick = jest.fn();
    render(
      <GeographySelectionButtons
        onIndividualButtonClick={onIndividualButtonClick}
        onPredefinedSelectionButtonClick={jest.fn()}
        geographySelection={GeographySelection.Individual}
        renderPreSelected={true}
      />
    );

    const individualButton = screen.getByText('Individual');
    individualButton.click();

    expect(onIndividualButtonClick).toHaveBeenCalledTimes(1);
  });

  it('should call onPredefinedSelectionButtonClick with the correct selection when predefined selection button is clicked', () => {
    const onPredefinedSelectionButtonClick = jest.fn();
    const { rerender } = render(
      <GeographySelectionButtons
        onIndividualButtonClick={jest.fn()}
        onPredefinedSelectionButtonClick={onPredefinedSelectionButtonClick}
        geographySelection={GeographySelection.Individual}
        renderPreSelected={true}
      />
    );

    const predefinedSelectionButtons = screen
      .getAllByRole('button')
      .filter((button) => {
        return button.textContent !== 'Individual';
      });

    for (let index = 0; index < predefinedSelectionButtons.length; index++) {
      const button = predefinedSelectionButtons[index];
      button.click();

      expect(onPredefinedSelectionButtonClick).toHaveBeenNthCalledWith(
        index + 1,
        Object.values(PredefinedSelection)[index]
      );

      rerender(
        <GeographySelectionButtons
          onIndividualButtonClick={jest.fn()}
          onPredefinedSelectionButtonClick={onPredefinedSelectionButtonClick}
          geographySelection={
            Object.values(PredefinedSelection)[
              index
            ] as unknown as GeographySelection
          }
          renderPreSelected={true}
        />
      );

      expect(
        screen.getAllByRole('button').filter((filteredButton) => {
          return filteredButton.textContent == button.textContent;
        })[0]
      ).toHaveClass(
        'rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-apian-yellow'
      );
    }
  });
});
