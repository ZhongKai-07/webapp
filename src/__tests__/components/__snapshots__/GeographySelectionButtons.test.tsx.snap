// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeographySelectionButtons should match the snapshot with preselected 1`] = `
<div>
  <div
    class="w-1/3justify-center flex gap-[1em] mt-[1em]"
  >
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-apian-yellow"
    >
      Individual
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50"
    >
      Isleofwight
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50"
    >
      Camden
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50"
    >
      Islington
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50"
    >
      Kensington
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50"
    >
      Northumbria
    </button>
  </div>
</div>
`;

exports[`GeographySelectionButtons should match the snapshot without preselected 1`] = `
<div>
  <div
    class="w-1/3justify-center flex gap-[1em] mt-[1em]"
  >
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-apian-yellow"
    >
      Individual
    </button>
  </div>
</div>
`;
